### PR TITLE
Deffered Object Store Config Parsing (#4191) 

### LINKED ISSUE #4191: Return Error on Invalid Config Value
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently if `AWS_CHECKSUM_ALGORITHM` is set to anything other than a capitilization of `sha256` `AmazonS3Builder::from_env` will panic.

Similarly non-truthy string values will be interpreted as `false`, which could lead to a confusing situation where a config value of "enabled" is interpreted as false.

**Describe the solution you'd like**

We should return an error for these

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4191

# Rationale for this change
 

This closes #4191 whilst also making me happier that the current pattern is workable to support things like durations which need fallible parsing, etc...

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/client/mod.rs [hunk 1]
```diff
@@ -43,6 +45,14 @@ fn map_client_error(e: reqwest::Error) -> super::Error {
...
     concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"),);
 
+/// Configuration keys for [`ClientOptions`]
+#[derive(PartialEq, Eq, Hash, Clone, Debug, Copy, Deserialize, Serialize)]
+#[non_exhaustive]
+pub enum ClientConfigKey {
+    /// Allow non-TLS, i.e. non-HTTPS connections
+    AllowHttp,
+}
+
```
Comment (roeap) [inline object_store/src/client/mod.rs:55]: Some users in delta-rs are asking to expose more client options on the python side, for this having some more keys defined might be helpful. Do you think this is something object_store wants to do in a follow-up?

**Review by roeap — APPROVED** [review]
really like this update!

File: object_store/src/client/mod.rs (hunk 1 above)
Comment (tustvold) [reply object_store/src/client/mod.rs:55]: > Do you think this is something object_store wants to do in a follow-up

Yes, 100%

Comment (tustvold) [reply object_store/src/client/mod.rs:55]: https://github.com/apache/arrow-rs/pull/4202 is the next step in exposing client options to the python side

**Pull request merged**